internal/graph/lr2streams.go

```diff
@@ -53,87 +60,53 @@ func runDispatchAndChecker(
+type checkAndDispatchRunner struct {
+	parentRequest     ValidatedLookupResources2Request
+	foundResources    dispatchableResourcesSubjectMap2
+	ci                cursorInformation
+	parentStream      dispatch.LookupResources2Stream
+	newSubjectType    *core.RelationReference
+	entrypoint        typesystem.ReachabilityEntrypoint
+	lrDispatcher      dispatch.LookupResources2
+	checkDispatcher   dispatch.Check
+	dispatchChunkSize uint16
+
 	filteredSubjectIDs []string
-	entrypoint         typesystem.ReachabilityEntrypoint
-	lrDispatcher       dispatch.LookupResources2
-	checkDispatcher    dispatch.Check
-	dispatchChunkSize  uint16
+	currentCheckIndex  int
 
 	taskrunner *taskrunner.TaskRunner
 
 	lock *sync.Mutex
 }
 
-func (rdc *dispatchAndCheckRunner) dispatchAndCollect(ctx context.Context, cursor *v1.Cursor) ([]*v1.DispatchLookupResources2Response, error) {
-	collectingStream := dispatch.NewCollectingDispatchStream[*v1.DispatchLookupResources2Response](ctx)
-	err := rdc.lrDispatcher.DispatchLookupResources2(&v1.DispatchLookupResources2Request{
-		ResourceRelation: rdc.parentRequest.ResourceRelation,
-		SubjectRelation:  rdc.newSubjectType,
-		SubjectIds:       rdc.filteredSubjectIDs,
-		TerminalSubject:  rdc.parentRequest.TerminalSubject,
-		Metadata: &v1.ResolverMeta{
-			AtRevision:     rdc.parentRequest.Revision.String(),
-			DepthRemaining: rdc.parentRequest.Metadata.DepthRemaining - 1,
-		},
-		OptionalCursor: cursor,
-		OptionalLimit:  uint32(rdc.dispatchChunkSize),
-	}, collectingStream)
-	return collectingStream.Results(), err
-}
-
-func (rdc *dispatchAndCheckRunner) runDispatch(ctx context.Context, cursor *v1.Cursor) error {
-	rdc.lock.Lock()
-	if rdc.ci.limits.hasExhaustedLimit() {
-		rdc.lock.Unlock()
-		return nil
-	}
-	rdc.lock.Unlock()
-
-	collected, err := rdc.dispatchAndCollect(ctx, cursor)
-	if err != nil {
-		return err
-	}
-
-	if len(collected) == 0 {
-		return nil
-	}
-
-	// Kick off a worker to filter the results via a check and then publish what was found.
+func (rdc *checkAndDispatchRunner) runAndWait() error {
+	// Kick off a check at the current cursor, to filter a portion of the initial results set.
 	rdc.taskrunner.Schedule(func(ctx context.Context) error {
-		return rdc.runChecker(ctx, collected)
+		return rdc.runChecker(ctx, rdc.currentCheckIndex)
 	})
 
-	// Start another dispatch at the cursor of the last response, to run in the background
-	// and collect more results for filtering while the checker is running.
-	rdc.taskrunner.Schedule(func(ctx context.Context) error {
-		return rdc.runDispatch(ctx, collected[len(collected)-1].AfterResponseCursor)
-	})
-
-	return nil
+	return rdc.taskrunner.Wait()
 }
 
-func (rdc *dispatchAndCheckRunner) runChecker(ctx context.Context, collected []*v1.DispatchLookupResources2Response) error {
+func (rdc *checkAndDispatchRunner) runChecker(ctx context.Context, startingIndex int) error {
 	rdc.lock.Lock()
 	if rdc.ci.limits.hasExhaustedLimit() {
```

Comment:
I think the limit should just be an atomic, and then you can get rid of the locking.

Reply:
Its not; we'd have to change the implementation and its use everywhere

Reply:
My suggestion was to change it to an atomic and update the callsites. That can be a followup though.